proposals/0032-middleware-api.md

```diff
@@ -0,0 +1,469 @@
+```js
+export const onRequest = (context, next) => {
+    if (!context.request.url.endsWith("/")) {
+        context.locals.isIndex = false;
+    } else {
+        context.locals.isIndex = true;
+    }
+}
+```
+
+Or, set some logic to make redirects:
+```js
+const redirects = new Map([
+    ["/old-1", "/new-1"],
+    ["/old-1", "/new-1"]
+])
+
+export const onRequest = (context, next) => {
+    for (const [oldRoute, newRoute] of redirects) {
+        if (context.request.url.endsWith(oldRoute)) {
+            return context.redirect(newRoute, 302);
+        }
+    }
+}
+```
+
+# Background & Motivation
+
+Middleware has been one of the most heavily requested feature in Astro. 
+It's useful for handling common tasks like auth guards and setting cache headers. 
+
+For me, it would make handling authentication much easier.
+
+
+# Goals
+
+- Provide a way intercept requests and responses, allowing users to set cookies and headers
+- Works both in SSR and SSG mode
+- Allow users to use community-created middlewares (libraries)
+  - Make available via integrations API.
+- Provide an API for request-specific data
+- Non-Node runtimes specific APIs. ie. Cloudflare Durable Objects.
+  - Add middleware from adapter.
+
+# Non-Goals
+
+- Route specific middleware, middlewares that are run **only on specific** routes
+
+# Detailed Design
+
+## Changes from the Stage 2 proposal
+
+- `resolve` has been renamed to `next`;
+- `next` doesn't accept an `APIContext` to work;
+- `locals` values need to be serializable to avoid the introduction of
+  non-user code from third-party libraries that can run scripts;
+- `middleware` export function has been renamed `onRequest`. This name change
+  has two benefits:
+  1. It shows intent and explains when this function is called;
+  2. It allows adding more functions with the `on*` prefix in the future, which could show the intent
+     of when the function is called in the Astro route life cycle;
+
+## Implementation instructions
+
+To define a middleware, a user must create a physical file under the `src/` folder, called `middleware.js`.
+
+The resolution of the file follows the ECMA standards, which means that the following
+alternatives are all valid in Astro:
+- `src/middleware.js`
+- `src/middleware.ts`
+- `src/middleware/index.js`
+- `src/middleware/index.ts`
+
+The file **must export** a function called `onRequest`. The exported function
+_must not be a **default** export_.
+
+Eventually, the file system of the `src/` folder will look like this:
+
+```
+src
+├── env.d.ts
+├── middleware
+│   └── index.ts
+└── pages
+    ├── first.astro
+    ├── index.astro
+    └── second.astro
+```
+
+Every time a page or endpoint is about to be rendered, the middleware is called.
+
+A middleware will look like this in TypeScript:
+
+```ts
+import { MiddlewareRequestHandler, APIContext, MiddlewareNextResponse } from "astro"
+
+const onRequest: MiddlewareRequestHandler = async (context: APIContext, next: MiddlewareNextResponse) => {
```

Comment:
Discussed on discord, but it could be nice to have a `defineMiddleware` export from `astro/middleware`. That would be like `defineConfig` in the Astro config. That function doesn't actually do anything other than make it so the what you pass to it is properly typed.